### ISSUE TITLE
Implement EncoderMemoryBridge and shared encoder-derived global KV

### ISSUE THREAD

**Issue by iamdarshg** [iamdarshg/KestrelVL-7B#4]
Parent epic: #2
Depends on: #3

## Goal

Implement the learned interface that turns Qwen3.5-2B encoder states into decoder-consumable shared global memory for Qwen3.5-9B.

Do **not** directly splice 2048-d encoder hidden states into 4096-d decoder residual states. The decoder should consume a dedicated external-memory representation.

## Initial design

```text
encoder final/intermediate states (d=2048)
        |
      RMSNorm
        |
   Linear 2048 -> 4096
        |
      SwiGLU
        |
  residual adapter
        |
   K/V projections
        |
shared global encoder memory
```

Add two optional source taps:

- **detail tap:** intermediate encoder representation for exact strings/code/numbers;
- **semantic tap:** final encoder representation for compressed semantic context.

## Functional requirements

- New decoder-memory contribution must be gated and initialized to zero/near-zero so source decoder behavior is initially recoverable.
- Bridge parameters are independently trainable/freezeable.
- Support one shared global memory representation with lightweight per-layer adaptation rather than storing independent full prompt KV for every decoder layer.
- Expose both dense-reference and sparse-memory interfaces so CSA2/indexer work can be validated independently.
- Memory object must have deterministic serialization/state semantics compatible with #1.
- Do not retain a second unbounded full-resolution prompt KV history in the decoder.

## Distillation targets

The bridge training API should support matching:

- teacher logits;
- selected decoder hidden states;
- attention outputs;
- teacher prompt-token relevance / memory retrieval targets;
- optional KV-space reconstruction where meaningful.

## Tests

- zero-gate equivalence to source decoder;
- shape/device/dtype coverage;
- deterministic prefill -> decode reuse;
- one-time encoder-memory construction;
- no prompt re-encoding during ordinary token decode;
- detail/semantic tap ablation;
- memory serialization round-trip.

## Acceptance criteria

- [ ] Encoder state is mapped cleanly from 2048-d to decoder memory space.
- [ ] Qwen3.5-9B can generate while conditioning on the mapped encoder memory.
- [ ] Disabling the memory path recovers the source decoder baseline.
- [ ] Shared memory is reusable across decoder layers/tokens.
- [ ] No independent full prompt KV copy is required for every decoder layer.
- [ ] Interface is ready for CSA2 + hierarchical sparse retrieval.

**Comment by iamdarshg** [comment]
## Batch 1 CPU evidence — acceptance criteria PASS (tiny doubles)

**Commit:** \8cad27b\. **Tests:** \	ests/test_encoder_bridge.py\ — 23/23 pass on CPU. Full suite 174 green.

**Proof per criterion:**
- Clean 2048→memory mapping: exact pipeline RMSNorm → Linear(2048→4096, no bias) → SwiGLU residual adapter (zero-init output scale) → K/V projections. No hidden-state splice exists — \splice_into_decoder_states\ always raises (tested).
- Conditioning path: \uild_semantic\ + optional \uild_detail\ → \uild_memory\ → typed \EncoderMemory\ (global_k/v + sha256 fingerprint + dtype/device tags). Detail=None works; mismatched seq lens fuse via broadcast-add (tested).
- Gate-disabled recovery: \output_gate\=exactly 0.0 at init → memory K/V exactly zero; \lank_memory_like\ exact zeros (tested). Feeds #3's zero-gate recovery (max_abs_diff=0.0).
- Shared reuse: \
euse_for_layers(n)\ returns views sharing parent \data_ptr\ (asserted) — no per-layer KV duplication. \uild_memory\ is pure/stateless; building twice accumulates nothing (tested).
- Serialization: \	o_dict/from_dict\ fingerprint-verified (tamper raises); \state_dict_snapshot/metadata/load_snapshot_strict\ rejects dim/dtype/device mismatch and missing keys (tested).
- Dtype/device: explicit ctor args, forward casts, outputs tagged, non-finite inputs raise (tested). Backward flows when unfrozen, blocked when frozen (tested).
- CSA2/indexer-ready: \dense_reference()\ + \sparse_memory_stub(top_k)\ consumed by #5/#6 (integration-tested: indexer recall 1.0 on bridge memory).
